universal-login-relayer/lib/services/CounterfactualIdentityService.js

```diff
@@ -0,0 +1,118 @@
+import Identity from 'universal-login-contracts/build/Identity';
+import {addressToBytes32} from '../utils/utils';
+import ethers, {utils, Wallet} from 'ethers';
+import defaultDeployOptions from '../config/defaultDeployOptions';
+
+
+class CounterfactualIdentityService {
+  constructor(wallet, ensService, authorisationService, hooks, provider) {
+    this.wallet = wallet;
+    this.abi = Identity.interface;
+    this.ensService = ensService;
+    this.authorisationService = authorisationService;
+    this.codec = new utils.AbiCoder();
+    this.hooks = hooks;
+    this.provider = provider;
+    this.deploymentValue = utils.bigNumberify(utils.parseEther('0.4'));
+  }
+
+  async create(managementKey, ensName, overrideOptions = {}) {
+    let deployTransaction = this.createDeployTransaction(managementKey, ensName);
+
+    deployTransaction = this.setupDeployTransaction(deployTransaction, overrideOptions);
+
+    const signedTransaction = this.signDeployTransaction(deployTransaction);
+
+    const counterfactualTransaction = this.counterfactualizeSignedTransaction(signedTransaction);
+
+    const fullCounterfactualData = await this.extractCounterfactualParams(counterfactualTransaction);
+
+    this.hooks.emit('AddressGenerated', {address: fullCounterfactualData.contractAddress});
```

Comment:
From this architecture it seems that it's complicated to be able to predict an address before generating it. Maybe we should add an optional "random nonce" that is used create different addresses, and an async function that predicts which address will be generated by that nonce? In this manner, if the app wants, it can allow the user to pick his address/icon

Reply:
Yes, you cannot pick your address. We can do the prediction function (working like "generate random" button that you can press multiple times). However, this will alter quite some the flow of creation. Alternatively, we can generate 10 accounts on the request, and let the user pick one of the 10 (via the first execution). Still requires changes to the flow though